File: apps/grokking-ml-fsharp/src/GrokkingMl/Ch12Ensembles.fs
/// 第 12 章: アンサンブル学習。
/// 弱い学習器を集めて強い学習器を作る。多数決（バギング）と
/// 逐次的な重み付け（AdaBoost）の 2 つの流儀を実装する。
module GrokkingMl.Ch12Ensembles

open System
open GrokkingMl.Ch09DecisionTrees

/// 復元抽出で元と同じ大きさの標本を作る。
let bootstrapSample (rng: Random) (points: Point list) (labels: int list) =
    let indices = List.init (List.length points) (fun _ -> rng.Next(List.length points))
    (indices |> List.map (fun i -> points[i]), indices |> List.map (fun i -> labels[i]))

/// 多数決で予測する木の集まり。
type Forest = { Trees: Tree list }

/// 各木の投票。
let votes (forest: Forest) (point: Point) =
    forest.Trees |> List.map (fun tree -> predict tree point)

/// 多数決による予測。
let forestPredict (forest: Forest) (point: Point) = majorityLabel (votes forest point)

/// バギング。復元抽出した標本ごとに木を育て、多数決で予測する。
let trainForestWith
    (treeCount: int)
    (maxDepth: int)
    (impurity: Impurity)
    (seed: int)
    (points: Point list)
    (labels: int list)
    =
    let rng = Random(seed)

    { Trees =
        List.init treeCount (fun _ ->
            let samplePoints, sampleLabels = bootstrapSample rng points labels
            buildTreeWith impurity maxDepth 1 samplePoints sampleLabels) }

/// 既定の設定（木 10 本、深さ 1、ジニ不純度、シード 0）でバギングする。
let trainForest (points: Point list) (labels: int list) =
    trainForestWith 10 1 giniImpurity 0 points labels

/// AdaBoost の弱学習器。発言権（Weight）を持つ。
type WeightedTree = { Tree: Tree; Weight: float }

/// 重み付き多数決で予測する学習器の列。ラベルは +1 / -1。
type AdaBoost = { Learners: WeightedTree list }

/// 重み付き投票のスコア。
let boostScore (model: AdaBoost) (point: Point) =
    model.Learners
    |> List.sumBy (fun learner -> learner.Weight * float (predict learner.Tree point))

/// AdaBoost による予測。
let boostPredict (model: AdaBoost) (point: Point) =
    if boostScore model point >= 0.0 then 1 else -1

/// 重み付き誤り率。重みの大きい点を間違えるほど大きくなる。
let weightedError (tree: Tree) (points: Point list) (labels: int list) (weights: float list) =
    let total = List.sum weights

    if total = 0.0 then
        0.0
    else
        List.zip3 points labels weights
        |> List.sumBy (fun (point, label, weight) -> if predict tree point <> label then weight else 0.0)
        |> fun wrong -> wrong / total

/// 弱学習器の発言権。誤り率が小さいほど大きい。
let learnerWeight (error: float) =
    let epsilon = 1e-10
    let clamped = error |> max epsilon |> min (1.0 - epsilon)
    0.5 * log ((1.0 - clamped) / clamped)

/// 重みを反映した木。重みに比例して点を複製してから学習する。
let buildTreeWithWeights
    (impurity: Impurity)
    (maxDepth: int)
    (points: Point list)
    (labels: int list)
    (weights: float list)
    =
    let scale = 100.0

    let replicated =
        List.zip3 points labels weights
        |> List.collect (fun (point, label, weight) ->
            let count = max 1 (int (Math.Round(weight * scale)))
            List.replicate count (point, label))

    let replicatedLabels = replicated |> List.map snd

    if (replicatedLabels |> List.distinct |> List.length) = 1 then
        Leaf(List.head replicatedLabels)
    else
        buildTreeWith impurity maxDepth 1 (replicated |> List.map fst) replicatedLabels

/// AdaBoost。間違えた点の重みを上げながら弱学習器を足していく。
let trainAdaBoostWith
    (rounds: int)
    (maxDepth: int)
    (impurity: Impurity)
    (points: Point list)
    (labels: int list)
    =
    let initialWeights = List.replicate (List.length points) (1.0 / float (List.length points))

    let step (learners, weights) _ =
        let tree = buildTreeWithWeights impurity maxDepth points labels weights
        let error = weightedError tree points labels weights

        // 当てずっぽう以下の学習器は採用しない
        if error >= 0.5 then
            (learners, weights)
        else
            let alpha = learnerWeight error

            let updated =
                List.zip3 points labels weights
                |> List.map (fun (point, label, weight) ->
                    weight * exp (-alpha * float label * float (predict tree point)))

            let total = List.sum updated
            (learners @ [ { Tree = tree; Weight = alpha } ], updated |> List.map (fun w -> w / total))

    let learners, _ = List.fold step ([], initialWeights) [ 1..rounds ]
    { Learners = learners }

/// 既定の設定（5 ラウンド、深さ 1、ジニ不純度）で学習する。
let trainAdaBoost (points: Point list) (labels: int list) =
    trainAdaBoostWith 5 1 giniImpurity points labels

/// 森の正解率。
let forestAccuracy (forest: Forest) (points: Point list) (labels: int list) =
    List.map2 (fun point label -> if forestPredict forest point = label then 1.0 else 0.0) points labels
    |> List.average

/// AdaBoost の正解率。
let boostAccuracy (model: AdaBoost) (points: Point list) (labels: int list) =
    List.map2 (fun point label -> if boostPredict model point = label then 1.0 else 0.0) points labels
    |> List.average

/// 1 本の木の正解率。
let treeAccuracy (tree: Tree) (points: Point list) (labels: int list) =
    List.map2 (fun point label -> if predict tree point = label then 1.0 else 0.0) points labels
    |> List.average

File: apps/grokking-ml-python/src/grokking_ml/ch12_ensembles.py
"""第 12 章: アンサンブル学習。

弱い学習器を集めて強い学習器を作る。多数決（バギング・ランダムフォレスト）と
逐次的な重み付け（AdaBoost）の 2 つの流儀を実装する。
"""

from __future__ import annotations

import math
import random
from collections.abc import Sequence
from dataclasses import dataclass

from grokking_ml.ch09_decision_trees import (
    Impurity,
    Leaf,
    Point,
    Tree,
    build_tree,
    gini_impurity,
    majority_label,
)


def bootstrap_sample(
    points: Sequence[Point],
    labels: Sequence[int],
    rng: random.Random,
) -> tuple[list[Point], list[int]]:
    """復元抽出で元と同じ大きさの標本を作る。"""
    indices = [rng.randrange(len(points)) for _ in range(len(points))]
    return [points[i] for i in indices], [labels[i] for i in indices]


@dataclass(frozen=True)
class Forest:
    """多数決で予測する木の集まり。"""

    trees: tuple[Tree, ...]

    def votes(self, point: Point) -> list[int]:
        return [tree.predict(point) for tree in self.trees]

    def predict(self, point: Point) -> int:
        return majority_label(self.votes(point))


def train_forest(
    points: Sequence[Point],
    labels: Sequence[int],
    tree_count: int = 10,
    max_depth: int = 1,
    impurity: Impurity = gini_impurity,
    seed: int = 0,
) -> Forest:
    """バギング。復元抽出した標本ごとに木を育て、多数決で予測する。"""
    rng = random.Random(seed)
    trees = []
    for _ in range(tree_count):
        sample_points, sample_labels = bootstrap_sample(points, labels, rng)
        trees.append(build_tree(sample_points, sample_labels, max_depth=max_depth, impurity=impurity))
    return Forest(trees=tuple(trees))


@dataclass(frozen=True)
class WeightedTree:
    """AdaBoost の弱学習器。発言権（weight）を持つ。"""

    tree: Tree
    weight: float


@dataclass(frozen=True)
class AdaBoost:
    """重み付き多数決で予測する学習器の列。ラベルは +1 / -1。"""

    learners: tuple[WeightedTree, ...]

    def score(self, point: Point) -> float:
        return sum(learner.weight * learner.tree.predict(point) for learner in self.learners)

    def predict(self, point: Point) -> int:
        return 1 if self.score(point) >= 0 else -1


def weighted_error(
    tree: Tree,
    points: Sequence[Point],
    labels: Sequence[int],
    weights: Sequence[float],
) -> float:
    """重み付き誤り率。重みの大きい点を間違えるほど大きくなる。"""
    total = sum(weights)
    if total == 0.0:
        return 0.0
    wrong = sum(
        weight
        for point, label, weight in zip(points, labels, weights)
        if tree.predict(point) != label
    )
    return wrong / total


def learner_weight(error: float) -> float:
    """弱学習器の発言権。誤り率が小さいほど大きい。"""
    epsilon = 1e-10
    clamped = min(max(error, epsilon), 1.0 - epsilon)
    return 0.5 * math.log((1.0 - clamped) / clamped)


def train_adaboost(
    points: Sequence[Point],
    labels: Sequence[int],
    rounds: int = 5,
    max_depth: int = 1,
    impurity: Impurity = gini_impurity,
) -> AdaBoost:
    """AdaBoost。間違えた点の重みを上げながら弱学習器を足していく。"""
    weights = [1.0 / len(points)] * len(points)
    learners: list[WeightedTree] = []
    for _ in range(rounds):
        tree = build_tree_with_weights(points, labels, weights, max_depth, impurity)
        error = weighted_error(tree, points, labels, weights)
        if error >= 0.5:
            # 当てずっぽう以下の学習器は採用しない
            break
        alpha = learner_weight(error)
        learners.append(WeightedTree(tree=tree, weight=alpha))
        weights = [
            weight * math.exp(-alpha * label * tree.predict(point))
            for point, label, weight in zip(points, labels, weights)
        ]
        total = sum(weights)
        weights = [weight / total for weight in weights]
    return AdaBoost(learners=tuple(learners))


def build_tree_with_weights(
    points: Sequence[Point],
    labels: Sequence[int],
    weights: Sequence[float],
    max_depth: int,
    impurity: Impurity,
) -> Tree:
    """重みを反映した木。重みに比例して点を複製してから学習する。"""
    replicated_points: list[Point] = []
    replicated_labels: list[int] = []
    scale = 100
    for point, label, weight in zip(points, labels, weights):
        count = max(1, round(weight * scale))
        replicated_points.extend([point] * count)
        replicated_labels.extend([label] * count)
    if len(set(replicated_labels)) == 1:
        return Leaf(replicated_labels[0])
    return build_tree(replicated_points, replicated_labels, max_depth=max_depth, impurity=impurity)


def accuracy(model: Forest | AdaBoost, points: Sequence[Point], labels: Sequence[int]) -> float:
    """正解率。"""
    correct = sum(1 for point, label in zip(points, labels) if model.predict(point) == label)
    return correct / len(points)


def tree_accuracy(tree: Tree, points: Sequence[Point], labels: Sequence[int]) -> float:
    """1 本の木の正解率。"""
    correct = sum(1 for point, label in zip(points, labels) if tree.predict(point) == label)
    return correct / len(points)
